src/backend/distributed/commands/schema_based_sharding.c

```diff
@@ -227,18 +643,212 @@ citus_internal_unregister_tenant_schema_globally(PG_FUNCTION_ARGS)
 	if (HeapTupleIsValid(namespaceTuple))
 	{
 		ReleaseSysCache(namespaceTuple);
-
 		ereport(ERROR, (errmsg("schema is expected to be already dropped "
 							   "because this function is only expected to "
 							   "be called from Citus drop hook")));
 	}
-
 	uint32 tenantSchemaColocationId = SchemaIdGetTenantColocationId(schemaId);
 
 	DeleteTenantSchemaLocally(schemaId);
 	SendCommandToWorkersWithMetadata(TenantSchemaDeleteCommand(schemaNameStr));
 
 	DeleteColocationGroup(tenantSchemaColocationId);
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * citus_schema_distribute gets a regular schema name, then converts it to a tenant
+ * schema.
+ */
+Datum
+citus_schema_distribute(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+	EnsureCoordinator();
+
+	Oid schemaId = PG_GETARG_OID(0);
+	EnsureSchemaOwner(schemaId);
+
+	/* Prevent concurrent schema deletion and table creation under the schema */
+	LockDatabaseObject(NamespaceRelationId, schemaId, 0, AccessExclusiveLock);
+
+	/*
+	 * It is possible that by the time we acquire the lock on schema,
+	 * concurrent DDL has removed it. We can test this by checking the
+	 * existence of schema.
+	 */
+	if (!SearchSysCacheExists1(NAMESPACEOID, ObjectIdGetDatum(schemaId)))
+	{
+		ereport(ERROR, (errcode(ERRCODE_UNDEFINED_SCHEMA),
+						errmsg("concurrent operation removed the schema")));
+	}
+
+	/* Return if the schema is already a tenant schema */
+	char *schemaName = get_namespace_name(schemaId);
+	if (IsTenantSchema(schemaId))
+	{
+		ereport(NOTICE, (errmsg("schema %s is already distributed", schemaName)));
+		PG_RETURN_VOID();
+	}
+
+	/* Collect all tables by their types in the schema */
+	SchemaTables *schemaTables = TablesUnderSchema(schemaId);
+
+	/* Makes sure the schema can be distributed. */
+	EnsureSchemaCanBeDistributed(schemaId, schemaTables);
+
+	/* We already verified we only have Citus local and Postgres local tables. */
+	List *tableIdList = GetAllTableIdListUnderSchema(schemaTables);
+
+	/* Convert each local relation to a single shard relation */
+	Oid relationId = InvalidOid;
+	List *tableListToConvert = NIL;
+	foreach_oid(relationId, tableIdList)
+	{
+		/*
+		 * Skip partitions as they would be distributed by the parent table.
+		 *
+		 * We should filter out partitions here before distributing the schema.
+		 * Otherwise, converted partitioned table would change oid of partitions and its
+		 * partition tables would fail with oid not exist.
+		 */
+		if (PartitionTable(relationId))
+		{
+			continue;
+		}
+
+		/*
+		 * Error early before distributing the schema if concurrent drop or some alter
+		 * statements occurred on the table. Note that we already had the AccessShareLock
+		 * lock when we call PartitionTable, so after this point we are guaranteed to
+		 * prevent concurrent drop and some modification on the table.
+		 */
+		EnsureRelationExists(relationId);
```

Comment:
We should do it like below:

```c
                LockRelationOid(relationId, AccessShareLock);
		EnsureRelationExists(relationId);

		if (PartitionTable(relationId))
```

Reply:
We may do a table owner check for the sake of early error. Otherwise, table check will be done by other methods during distribute/undistribute.

Reply:
If the user has enough privileges for some of the tables in the schema but doesn't have for some other, deferring the ownership check would cause blocking the writes to prior ones unncessarily.

Given that it's a one liner / cheap check that could save us from such situations, let's this one as well.